### PR TITLE
Deprecate Broker::auto_publish()

### PR DESCRIPTION
**Pull request by awelzel**
This PR removes all `Broker::auto_publish()` usages in base and policy scripts and replaces them with explicit `Broker::publish()` calls. In the future, these can be replaced with 's/Broker::publish/Cluster::publish/g`.

This fixes `Broker::publish(..., table())` not working quite right as that was needed for sumstats.

I think I'm still on the side that even if it makes some places more verbose, I find the explicitness immediately helps with the question "is this intended to be handled locally or on a remote node?", even if some places got a bit explicit.

#3637 as background.

I've marked Broker::auto_publish() deprecated for v8.1. I could see making this 9.1, too.

### CODE REVIEW

**Comment by awelzel** [comment]
@J-Gras - there's the `remote_event_auto_ts.zeek` test case which tests that remote events use the "intended timestamp" for scheduled events that are also being "auto published".

I think a future `publish_with_ts()` or `publish_with_metadata()` could work here. It seems a cool trick that one can control the timestamp this way, but `auto_publish()` and using the "intended timestamp" via `schedule` isn't very obvious.

 I looked at some of your code and it doesn't seem you rely on this. Is that correct? Did you plan on?

**Comment by J-Gras** [comment]
> @J-Gras - there's the `remote_event_auto_ts.zeek` test case which tests that remote events use the "intended timestamp" for scheduled events that are also being "auto published".

Yep, I believe I added this, because auto publishing is using a different code path and I had to make sure to add the metadata to broker messages on that path as well.

> I think a future `publish_with_ts()` or `publish_with_metadata()` could work here. It seems a cool trick that one can control the timestamp this way, but `auto_publish()` and using the "intended timestamp" via `schedule` isn't very obvious.

I don't follow here: Every published event should have the appropriate metadata set. In this case the current (network) timestamp. Introducing a `publish_with_ts()` sounds like 1) the default publish wouldn't carry the timestamp or 2) the user could manually set the timestamp. Both could kill my use case. So I'd just stick with `publish` that always carries the metadata.

> I looked at some of your code and it doesn't seem you rely on this. Is that correct? Did you plan on?

Not sure why the metadata would be different than the others? If I am not mistaken, we can just add handlers for the scheduled events and explicitly call publish from these events :)

**Comment by awelzel** [comment]
> Every published event should have the appropriate metadata set. In this case the current (network) timestamp.

Correct, but for a scheduled event it's actually the "intended time" that's used, which might be earlier than whatever is network time (think scheduling an event in 5sec, but the next packet arriving only 60 seconds into the future). The scheduled "auto_publish" will use the 5second timestamp.

I did setup a call with you for tomorrow to go through this.

File: testing/btest/scripts/base/frameworks/cluster/start-it-up-logger.zeek [hunk 1]
```diff
@@ -51,15 +51,6 @@ event fully_connected(n: string)
 			terminate();
 			}
 		}
-	else
-		{
-		print "sent fully_connected event";
```
Comment (ckreibich) [inline testing/btest/scripts/base/frameworks/cluster/start-it-up-logger.zeek:56]: Hah! Could it get any more implicit ... I'm glad this is gone.

**Review by ckreibich — APPROVED** [review]
LGTM module minor nits that you're welcome to handle as you see fit. I much prefer the explicit approach throughout. The code is just nicer.

This took pretty long to review because checking the intended implicit behavior of the old auto-publishes is fairly subtle.

(Btw I'm struggling with the size of these PRs so if you could keep stuff like the event timing in separate ones, it'd help me a lot, thx!)

File: scripts/base/frameworks/notice/main.zeek [hunk 2]
```diff
@@ -536,38 +524,34 @@ hook Notice::notice(n: Notice::Info) &priority=-5
 		event Notice::begin_suppression(n$ts, n$suppress_for, n$note, n$identifier);
 		suppressing[n$note, n$identifier] = n$ts + n$suppress_for;
 @if ( Cluster::is_enabled() && Cluster::local_node_type() != Cluster::MANAGER )
-		event Notice::manager_begin_suppression(n$ts, n$suppress_for, n$note, n$identifier);
+		# Notify the manager about the new suppression, it'll broadcast
+		# to the other nodes in the cluster.
+		# Once we have global pub/sub, we could also unconditionally
+		# send to a notice specific topic for communicating
+		# suppressions directly to all nodes.
+		local e = Broker::make_event(Notice::begin_suppression,
```
[the remark this answers is not in the document]
Comment (awelzel) [reply scripts/base/frameworks/notice/main.zeek:532]: Not really, changing to Broker::publish() directly.

File: scripts/base/frameworks/notice/main.zeek [hunk 3]
```diff
@@ -536,38 +524,34 @@ hook Notice::notice(n: Notice::Info) &priority=-5
...
-	Broker::auto_publish(Cluster::worker_topic, Notice::begin_suppression);
-	Broker::auto_publish(Cluster::proxy_topic, Notice::begin_suppression);
-	}
-
-event Notice::manager_begin_suppression(ts: time, suppress_for: interval, note: Type,
-								identifier: string)
+event Notice::begin_suppression(ts: time, suppress_for: interval, note: Type, identifier: string)
 	{
-	event Notice::begin_suppression(ts, suppress_for, note, identifier);
+	local e = Broker::make_event(Notice::begin_suppression, ts, suppress_for, note, identifier);
```
[the remark this answers is not in the document]
Comment (awelzel) [reply scripts/base/frameworks/notice/main.zeek:544]: I had pushed back on #3637 as you might unknowingly re-publish `c:  connection` or so. If there's a ton of these coming up, sure. But I kind of hope this re-publishing goes away once there's a notion of global pub/sub established.

File: scripts/base/protocols/dhcp/main.zeek [hunk 4]
```diff
@@ -307,7 +300,12 @@ event DHCP::aggregate_msgs(ts: time, id: conn_id, uid: string, is_orig: bool, ms
 # Aggregate DHCP messages to the manager.
 event dhcp_message(c: connection, is_orig: bool, msg: DHCP::Msg, options: DHCP::Options) &priority=-5
 	{
-	event DHCP::aggregate_msgs(network_time(), c$id, c$uid, is_orig, msg, options);
+@if ( Cluster::is_enabled() && Cluster::local_node_type() != Cluster::MANAGER )
+		Broker::publish(Cluster::worker_topic, DHCP::aggregate_msgs,
```
Comment (awelzel) [inline scripts/base/protocols/dhcp/main.zeek:304]: There's a bug, this needs to be `manager_topic` :-(

File: scripts/base/frameworks/sumstats/cluster.zeek [hunk 5]
```diff
@@ -154,16 +151,20 @@ event SumStats::cluster_get_result(uid: string, ss_name: string, key: Key, clean
 		{
 		if ( uid in sending_results && key in sending_results[uid] )
 			{
+			# XXX: Is that comment stale?
```
[the remark this answers is not in the document]
Comment (awelzel) [reply scripts/base/frameworks/sumstats/cluster.zeek]: I dropped it. I can see that *maybe* auto publish would behave funky, but an explicit Broker::publish() will serialize directly.

[the remark this answers is not in the document]
Comment (0xxon) [reply scripts/base/frameworks/sumstats/cluster.zeek]: FWIW - I think this was a workaround for the fact that the pre-broker communication framework did not, by default, deep copy records. Instead if kept track of the pointers of both sides of inside elements. I assume all of this has not been necessary for a long time.

**Pull request merged**